velox/exec/TopNRowNumber.cpp

```diff
@@ -141,13 +239,41 @@ void TopNRowNumber::processInputRow(
 void TopNRowNumber::noMoreInput() {
   Operator::noMoreInput();
 
-  auto rowSize = data_->estimateRowSize();
+  updateEstimatedOutputRowSize();
+
+  outputBatchSize_ = outputBatchRows(estimatedOutputRowSize_);
+
+  if (spiller_ != nullptr) {
+    // Spill remaining data to avoid running out of memory while sort-merging
+    // spilled data.
+    spill();
+
+    spiller_->finishSpill();
+    recordSpillStats(spiller_->stats());
+
+    merge_ = spiller_->startMerge(0);
+  } else {
+    outputRows_.resize(outputBatchSize_);
+  }
+}
+
+void TopNRowNumber::updateEstimatedOutputRowSize() {
+  const auto optionalRowSize = data_->estimateRowSize();
+  if (!optionalRowSize.has_value()) {
+    return;
+  }
+
+  auto rowSize = optionalRowSize.value();
+
   if (rowSize && generateRowNumber_) {
-    rowSize.value() += sizeof(int64_t);
+    rowSize += sizeof(int64_t);
   }
 
-  outputBatchSize_ = outputBatchRows(rowSize);
-  outputRows_.resize(outputBatchSize_);
+  if (!estimatedOutputRowSize_.has_value()) {
```

Comment:
We take the max as the estimation to be conservative on output batch processing?

Reply:
This is how I was thinking about it. Being conservative. Let me know if you think a different strategy would be better.